File: LRD Cache/LRDcache.py
from collections import OrderedDict

class LRUCache:
    def __init__(self, capacity: int):
        self.cache = OrderedDict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        else:
            self.cache.move_to_end(key)  # Move to end to show it was recently used
            return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
        elif len(self.cache) >= self.capacity:
            self.cache.popitem(last=False)  # Pop the first item (least recently used)
        self.cache[key] = value

# Example usage
def test_LRUCache(operations, values):
    result = []
    cache = None
    for op, value in zip(operations, values):
        if op == "LRUCache":
            cache = LRUCache(value[0])
            result.append(None)
        elif op == "put":
            cache.put(value[0], value[1])
            result.append(None)
        elif op == "get":
            result.append(cache.get(value[0]))
    return result

operations = ["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
values = [[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
print(test_LRUCache(operations, values))  # Output: [null, null, null, 1, null, -1, null, -1, 3, 4]

File: LRD Cache/LRDcache.cpp
#include <iostream>
#include <unordered_map>
#include <list>

class LRUCache {
private:
    int capacity;
    std::list<std::pair<int, int>> cacheList;  // doubly linked list to store key-value pairs
    std::unordered_map<int, std::list<std::pair<int, int>>::iterator> cache;  // map from keys to list iterators

    void moveToFront(int key, int value) {
        // Remove the old entry from the list
        cacheList.erase(cache[key]);
        // Insert the updated entry at the front
        cacheList.push_front(std::make_pair(key, value));
        // Update the map to point to the new location
        cache[key] = cacheList.begin();
    }

public:
    LRUCache(int capacity) : capacity(capacity) {}

    int get(int key) {
        if (cache.find(key) == cache.end()) return -1;  // key not found
        auto value = cache[key]->second;
        moveToFront(key, value);
        return value;
    }

    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            // Key already exists, update the value and move to front
            moveToFront(key, value);
        } else {
            // Check if we have reached the capacity
            if (cacheList.size() == capacity) {
                // Remove the least recently used item from cache and list
                auto lru = cacheList.back();
                cache.erase(lru.first);
                cacheList.pop_back();
            }
            // Insert the new item at the front
            cacheList.push_front(std::make_pair(key, value));
            cache[key] = cacheList.begin();
        }
    }
};

int main() {
    LRUCache cache(2);
    cache.put(1, 1);
    cache.put(2, 2);
    std::cout << cache.get(1) << std::endl;  // prints 1
    cache.put(3, 3);                         // evicts key 2
    std::cout << cache.get(2) << std::endl;  // prints -1 (not found)
    cache.put(4, 4);                         // evicts key 1
    std::cout << cache.get(1) << std::endl;  // prints -1 (not found)
    std::cout << cache.get(3) << std::endl;  // prints 3
    std::cout << cache.get(4) << std::endl;  // prints 4
    return 0;
}
